### DISCUSSION TITLE
Quick custom kernel registration / speeding up makefile compile times

### DISCUSSION THREAD

**Discussion by EdSheltonUKAEA** [idaholab/moose#23637] [Q&A General] [answered]
Is there a quick way to register a custom kernel?  
The only way I know at the moment is to run a complete compile of MOOSE with these commands:
`cd ~/projects/YourAppName
make clobberall
make -j4
./run_tests -j4`

For me this takes about an hour each time.  So as a separate issue, I'm sure this is too slow.  How can I speed up the compile process when running the make file?
I've attached a copy of my makefile (renamed as a .txt file to enable me to attach it).
[Makefile.txt](https://github.com/idaholab/moose/files/10910768/Makefile.txt)

**Comment by GiudGiud** [comment]
Hello

You should be able to add a file and recompile only the kernels file. Make in the MOOSE ecosystem realizes what objects have already been compiled. 

My compiles take under 20s most of the time when I've modified or added only a single file. 

Guillaume

**Reply by EdSheltonUKAEA** [reply]
Thanks - how do I do this?  Do I need to edit my make file?
My MOOSE ecosystem obviously isn't doing what you describe.

**Reply by GiudGiud** [reply]
Are you doing `make clobberall` before making?

You do not need to edit the Makefile.

If you add a space in "Diffusion.C" then make again, does it compile everything?

**Comment by mfitzka** [comment, marked as answer]
In my experience it is normally not necessary to do a "make clobberall" before recompiling to add new custom kernels. With this command you basically rebuild everything in your app folder from the ground up instead of simply adding your new kernels.

If you have more CPU kernels available you can substitute the 4 in "make -j4" with as many as you have, which can increase speed but will in the end be a limitation by your hardware.

**Reply by GiudGiud** [reply]
no you should not have to do a make clobberall.
Make clobberall is only necessary if you straight up update MOOSE or libmesh, and even then it s not always the case that it s needed, you can sometimes "wing it"

**Reply by EdSheltonUKAEA** [reply]
I think that's it.  I've always done "clobberall" before "make", as I've only ever previously used "make" after a MOOSE update.

But what seems odd is that when doing a full make for an update, I get lines for kernels in modules which are set to "NO" in my makefile.

**Reply by GiudGiud** [reply]
So you get other modules by inheritance. 

The classic one that s unexpected is getting ray tracing because you included heat conduction and it relies on ray tracing for some radiative problems

**Reply by EdSheltonUKAEA** [reply]
Ah that all makes sense now.  Thanks!
